src/main/java/supercoder79/survivalgames/game/map/SurvivalGamesChunkGenerator.java

```diff
@@ -58,268 +58,290 @@
-					}
-				}
-			}
-		}
-	}
-
-	@Override
-	public void populateBiomes(Registry<Biome> registry, Chunk chunk) {
-		ChunkPos chunkPos = chunk.getPos();
-		((ProtoChunk)chunk).setBiomes(new BiomeArray(registry, chunkPos, this.biomeSource));
-	}
-
-	@Override
-	public void carve(long seed, BiomeAccess access, Chunk chunk, GenerationStep.Carver carver) {
-	}
-
-
-}
+        this.piecesByChunk = new Long2ObjectOpenHashMap<>();
+        this.piecesByChunk.defaultReturnValue(ImmutableList.of());
+
+        List<SurvivalGamesJigsawGenerator> generators = new ArrayList<>();
+
+        this.noiseGenerator.initialize(random, config);
+
+        ChunkMask mask = new ChunkMask();
+        ChunkBox townArea = new ChunkBox();
+
+        if (config.townDepth > 0) {
+            SurvivalGamesJigsawGenerator generator = new SurvivalGamesJigsawGenerator(server, this, piecesByChunk);
+            generator.arrangePieces(new BlockPos(0, 64, 0), new Identifier("survivalgames", "starts"), config.townDepth);
+            townArea = generator.getBox();
+            generators.add(generator);
+            mask.and(townArea);
+        }
+
+        for (int i = 0; i < config.outskirtsBuildingCount; i++) {
+            int startX = random.nextInt(config.borderConfig.startSize / 2) - random.nextInt(config.borderConfig.startSize / 2);
+            int startZ = random.nextInt(config.borderConfig.startSize / 2) - random.nextInt(config.borderConfig.startSize / 2);
+
+            if (!this.noiseGenerator.shouldOutskirtsSpawn(startX, startZ)) {
+                continue;
+            }
+
+            BlockPos start = new BlockPos(startX, 0, startZ);
+            ChunkPos chunkPos = new ChunkPos(start);
+
+            if (mask.isIn(chunkPos)) {
+                continue;
+            }
+
+            SurvivalGamesJigsawGenerator outskirtGenerator = new SurvivalGamesJigsawGenerator(server, this, piecesByChunk);
+            outskirtGenerator.arrangePieces(start, new Identifier("survivalgames", "outskirts_building"), 0);
+
+            mask.and(chunkPos);
+
+            generators.add(outskirtGenerator);
+        }
+        for (int i = 0; i < config.netherTownStructures; i++) {
+            int startX = random.nextInt(config.borderConfig.startSize / 2) - random.nextInt(config.borderConfig.startSize / 2);
+            int startZ = random.nextInt(config.borderConfig.startSize / 2) - random.nextInt(config.borderConfig.startSize / 2);
+
+            if (!this.noiseGenerator.shouldOutskirtsSpawn(startX, startZ)) {
+                continue;
+            }
+
+            BlockPos start = new BlockPos(startX, 0, startZ);
+            ChunkPos chunkPos = new ChunkPos(start);
+
+            if (mask.isIn(chunkPos)) {
+                continue;
+            }
+
+            SurvivalGamesJigsawGenerator outskirtGenerator = new SurvivalGamesJigsawGenerator(server, this, piecesByChunk);
+            outskirtGenerator.arrangePieces(start, new Identifier("survivalgames", "nether_outskirts"), 0);
```

Comment:
This should go into the config, so you can specify what type of outskirts you want